src/main/java/com/epam/eco/commons/kafka/helpers/CachedTopicOffsetRangeFetcher.java

```diff
@@ -0,0 +1,202 @@
+import org.slf4j.LoggerFactory;
+
+import com.epam.eco.commons.kafka.KafkaUtils;
+import com.epam.eco.commons.kafka.OffsetRange;
+
+import static java.time.OffsetDateTime.now;
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+import static org.apache.commons.lang3.Validate.notNull;
+
+/**
+ * @author Andrei_Tytsik
+ */
+public class CachedTopicOffsetRangeFetcher extends TopicOffsetRangeFetcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CachedTopicOffsetRangeFetcher.class);
+    private static final long EXPIRATION_TIME_MIN = 10L;
+    private static final Map<TopicPartition, OffsetRangeCacheRecord> offsetRangeCache = new ConcurrentHashMap<>();
+    private final static ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    private final static Lock readLock = lock.readLock();
+    private final static Lock writeLock = lock.writeLock();
+
+    private CachedTopicOffsetRangeFetcher(String bootstrapServers, Map<String, Object> consumerConfig) {
+        super(bootstrapServers,consumerConfig);
+    }
+
+    public static CachedTopicOffsetRangeFetcher with(Map<String, Object> consumerConfig) {
+        return new CachedTopicOffsetRangeFetcher(null, consumerConfig);
+    }
+
+    public static CachedTopicOffsetRangeFetcher with(String bootstrapServers) {
+        return new CachedTopicOffsetRangeFetcher(bootstrapServers, null);
+    }
+
+
+    public Map<TopicPartition, OffsetRange> fetchForPartitions(Collection<TopicPartition> partitions) {
+        Validate.notEmpty(partitions, "Collection of partitions is null or empty");
+        Validate.noNullElements(partitions, "Collection of partitions contains null elements");
+
+        try (KafkaConsumer<?, ?> consumer = new KafkaConsumer<>(consumerConfig)) {
+            return calculateIfAbsent(consumer, partitions);
+        }
+    }
+
+    public Map<TopicPartition, OffsetRange> fetchForTopics(String ... topicNames) {
+        return fetchForTopics(
+                topicNames != null ? Arrays.asList(topicNames) : null);
+    }
+
+    public Map<TopicPartition, OffsetRange> fetchForTopics(Collection<String> topicNames) {
+        Validate.notEmpty(topicNames, "Collection of topic names is null or empty");
+        Validate.noNullElements(topicNames, "Collection of topic names contains null elements");
+
+        try (KafkaConsumer<?, ?> consumer = new KafkaConsumer<>(consumerConfig)) {
+            List<TopicPartition> partitions =
+                    KafkaUtils.getTopicPartitionsAsList(consumer, topicNames);
+            return calculateIfAbsent(consumer, partitions);
+        }
+    }
+
+    Map<TopicPartition, OffsetRange> calculateIfAbsent(Consumer<?,?> consumer,
+                                                       Collection<TopicPartition> partitions) {
+
+        Map<TopicPartition, OffsetRange> results = partitions.stream()
+                .filter(topicPartition->nonNull(getCacheValue(topicPartition)))
```

Comment:
This filter cannot guarantee that next getCacheValue() call will return not null